File: mpp-proiect-java-culbec/MotorcycleContest/Model/src/main/java/Model/Race.java
package Model;

import java.util.Objects;

public class Race extends Entity<Integer> {
    private Integer engineCapacity = null;
    private Integer noParticipants = null;

    private Race() {
        super();
    }

    public static class Builder {
        public Builder() {

        }

        private Race member = new Race();

        public Builder reset() {
            this.member = new Race();
            return this;
        }

        public Builder setId(Integer id) {
            this.member.setId(id);
            return this;
        }

        public Builder setNoParticipants(Integer noParticipants) {
            this.member.noParticipants = noParticipants;
            return this;
        }

        public Builder setEngineCapacity(Integer engineCapacity) {
            this.member.engineCapacity = engineCapacity;
            return this;
        }

        public Race build() {
            return member;
        }
    }

    public Integer getEngineCapacity() {
        return engineCapacity;
    }

    public Integer getNoParticipants() {
        return noParticipants;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Race race = (Race) o;
        return Objects.equals(noParticipants, race.noParticipants) && Objects.equals(engineCapacity, race.engineCapacity);
    }

    @Override
    public int hashCode() {
        return Objects.hash(noParticipants, engineCapacity);
    }

    @Override
    public String toString() {
        return "Model.Race{" +
                "noParticipants=" + noParticipants +
                ", engineCapacity=" + engineCapacity +
                '}';
    }
}

File: mpp-proiect-csharp-culbec/Model/Race.cs
namespace Model;

public class Race : Entity<int>
{
    public int EngineCapacity { get; set; }
    public int NoParticipants { get; set; }

    private Race()
    {
    }

    public new class Builder
    {
        private Race _member = new Race();

        public Builder Reset()
        {
            _member = new Race();
            return this;
        }

        public Builder SetId(int id)
        {
            _member.Id = id;
            return this;
        }

        public Builder SetEngineCapacity(int engineCapacity)
        {
            _member.EngineCapacity = engineCapacity;
            return this;
        }

        public Builder SetNoParticipants(int noParticipants)
        {
            _member.NoParticipants = noParticipants;
            return this;
        }

        public Race Build()
        {
            return _member;
        }
    }

    public new static Builder NewBuilder()
    {
        return new Builder();
    }

    public override bool Equals(object? obj)
    {
        if (ReferenceEquals(null, obj)) return false;
        if (ReferenceEquals(this, obj)) return true;
        return base.Equals(obj) && obj.GetType() == GetType() && Equals(EngineCapacity, ((Race)obj).EngineCapacity) &&
               Equals(NoParticipants, ((Race)obj).NoParticipants);
    }

    public override int GetHashCode()
    {
        return base.GetHashCode() + EqualityComparer<int>.Default.GetHashCode(EngineCapacity) +
               EqualityComparer<int>.Default.GetHashCode(NoParticipants);
    }

    public override string ToString()
    {
        return
            $"{base.ToString()}, {nameof(EngineCapacity)}: {EngineCapacity}, {nameof(NoParticipants)}: {NoParticipants}";
    }
}